components/logins/android/src/main/java/mozilla/appservices/logins/MemoryLoginsStorage.kt

```diff
@@ -271,12 +271,62 @@ class MemoryLoginsStorage(private var list: List<ServerPassword>) : AutoCloseabl
                     "Invalid login: Neither `formSubmitUrl` and `httpRealm` are present",
                     InvalidLoginReason.NO_TARGET)
         }
+
+        checkIllegalValues(login)
     }
 
     override fun ensureValid(login: ServerPassword) {
         checkValidWithNoDupes(login)
     }
 
+    @Suppress("ThrowsCount")
+    private fun checkIllegalValues(login: ServerPassword) {
+        val fieldData = listOf(
+            Pair("formSubmitUrl", login.formSubmitURL),
+            Pair("httpRealm", login.httpRealm),
+            Pair("hostname", login.hostname),
+            Pair("usernameField", login.usernameField),
+            Pair("passwordField", login.passwordField),
+            Pair("username", login.username),
+            Pair("password", login.password)
+        )
+
+        fieldData.forEach {
+            if (it.second.toString().contains("\u0000")) {
+                throw InvalidRecordException(
+                    "Invalid login: Login has illegal field: `$it.first` contains Nul",
+                    InvalidLoginReason.ILLEGAL_FIELD_VALUE)
+            }
+        }
+
+        // Newlines are invalid in Desktop with the exception of the username and password fields.
+        val fieldDataSubset = fieldData.subList(0, fieldData.size - 2)
+
+        fieldDataSubset.forEach {
+            if (it.second.toString().contains("\n") || it.second.toString().contains("\r")) {
+                throw InvalidRecordException(
+                    "Invalid login: Login has illegal field: `$it.first` contains newline",
+                    InvalidLoginReason.ILLEGAL_FIELD_VALUE)
+            }
+        }
+
+        if (login.usernameField == ".") {
+            throw InvalidRecordException(
+                    "Invalid login: Login has illegal field: `usernameField` is a period",
+                    InvalidLoginReason.ILLEGAL_FIELD_VALUE)
+        }
+        if (login.formSubmitURL == ".") {
+            throw InvalidRecordException(
+                    "Invalid login: Login has illegal field: `formSubmitUrl` is a period",
+                    InvalidLoginReason.ILLEGAL_FIELD_VALUE)
+        }
+        if (login.hostname.contains(" (")) {
+            throw InvalidRecordException(
+                    "Invalid login: Login has illegal field: Origin is Malformed",
+                    InvalidLoginReason.ILLEGAL_FIELD_VALUE)
+        }
+    }
```

Comment:
@rfk I hope this is possible. I felt really bad duplicating this logic, particularly because it deviates from the original rust code a tad to appease the android linter.

Reply:
> Oh. I'm sad you had to duplicate all of this logic into here, and I wonder if we still need this separate implementation. The only consumer I could find from some quick `rg`ing around was the Lockwise android app, where is seems to be used for testing. I wonder if we could replace this with e.g. an in-memory or tempfile-based SQLite DB. @thomcc thoughts?
> 
> (That would be for a separate PR in any case)

@rfk I hope this is possible. I felt really bad duplicating this logic, particularly because it deviates from the original rust code a tad to appease the android linter.

Reply:
@rfk the goal is so that this could be used in cases where calling into Rust wasn't possible. This effectively precludes any implementation other than completely duplicating the Rust code in Kotlin.

That said I'd love to see it go, I've never liked it, and really understand why we were providing it in the first place (especially since the other products just mock things on their own usually, and you can call driectly into the rust now).

I also imagine it's not 100% compatible with the real impl (probably merely in the high-90s). Moreover, it forces us to have to have an interface defining things, which complicates our API compatibility story (every addition is a breaking change), as well as complicating the android-components mirror of this API (slightly). And, as you mention, much of its functionality can be emulated by using an in-memory DB, which can be done by passing ":memory:" as the path to the database (https://www.sqlite.org/inmemorydb.html).

That said, it shouldn't be done as part of this PR.

Reply:
I filed https://github.com/mozilla/application-services/issues/2310 to follow up.